modules/fvad/fvad.c

```diff
@@ -0,0 +1,348 @@
+	struct filter_arg fa = { au, NULL };
+	(void)ctx;
+
+	if (!stp || !af || !prm)
+		return EINVAL;
+
+	if (*stp)
+		return 0;
+
+	if (prm->ch != 1)
+		return EINVAL;
+
+	st = mem_zalloc(sizeof(*st), enc_destructor);
+	if (!st)
+		return ENOMEM;
+
+	st->fvad = fvad_new();
+	if (!st->fvad) {
+		mem_deref(st);
+		return ENOMEM;
+	}
+
+	int err = fvad_set_sample_rate(st->fvad, prm->srate);
+	if (err < 0) {
+		mem_deref(st);
+		return EINVAL;
+	}
+
+	uag_filter_calls(find_first_call, find_call, &fa);
+
+	*stp = (struct aufilt_enc_st *)st;
+
+	st->call = fa.call;
+
+	return 0;
+}
+
+
+static int decode_update(struct aufilt_dec_st **stp, void **ctx,
+			 const struct aufilt *af, struct aufilt_prm *prm,
+			 const struct audio *au)
+{
+	struct vad_dec *st;
+	struct filter_arg fa = { au, NULL };
+	(void)ctx;
+
+	if (!stp || !af || !prm)
+		return EINVAL;
+
+	if (*stp)
+		return 0;
+
+	if (prm->ch != 1)
+		return EINVAL;
+
+	st = mem_zalloc(sizeof(*st), dec_destructor);
+	if (!st)
+		return ENOMEM;
+
+	st->fvad = fvad_new();
+	if (!st->fvad) {
+		mem_deref(st);
+		return ENOMEM;
+	}
+
+	int err = fvad_set_sample_rate(st->fvad, prm->srate);
+	if (err < 0) {
+		mem_deref(st);
+		return EINVAL;
+	}
+
+	uag_filter_calls(find_first_call, find_call, &fa);
+
+	st->call = fa.call;
+
+	*stp = (struct aufilt_dec_st *)st;
+
+	return 0;
+}
+
+
+static bool auframe_vad(Fvad *fvad, struct auframe *af)
+{
+	static int chunk_times_ms[] = { 30, 20, 10 };
+
+	int16_t *buf = NULL;
+	int16_t *allocated = NULL;
+	bool detected = false;
+
+	/* convert to 16 bit linear */
+	if (af->fmt == AUFMT_S16LE) {
+		buf = af->sampv;
+	}
+	else {
+		allocated = (int16_t*)mem_alloc(sizeof(int16_t) * af->sampc,
+			NULL);
+		buf = allocated;
+		auconv_to_s16(buf, af->fmt, af->sampv, af->sampc);
+	}
+
+	size_t ms = af->sampc * 1000 / af->srate;
+	size_t pos = 0;
+
+	/* process all chunk_sizes that fvad accepts */
+	for (size_t chunk_time_index = 0;
+		chunk_time_index < RE_ARRAY_SIZE(chunk_times_ms);
+		++chunk_time_index) {
+
+		const size_t chunk_time = chunk_times_ms[chunk_time_index];
+		const size_t sampc = af->srate / 1000 * chunk_time;
+
+		for (size_t i = 0; i < ms / chunk_time; ++i) {
+			int err = fvad_process(fvad, buf + pos, sampc);
+			pos += sampc;
+			ms -= chunk_time;
+			if (err > 0) {
+				detected = true;
+				goto out;
+			}
+			else if (err < 0) {
+				warning("fvad: fvad_process(%d) failed\n",
+					sampc);
+				goto out;
+			}
+		}
+	}
+
+	if (pos != af->sampc) {
+		warning("fvad: fvad_process: samples left over: %d\n",
+			af->sampc - pos);
+	}
+
+out:
+	mem_deref(allocated);
+
+	return detected;
+}
+
+
+static int encode(struct aufilt_enc_st *st, struct auframe *af)
+{
+	struct vad_enc *vad = (void *)st;
+
+	if (!st || !af)
+		return EINVAL;
+
+	bool vad_tx = auframe_vad(vad->fvad, af);
+
+	if (vad_tx != vad->vad_tx) {
+		vad->vad_tx = vad_tx;
+
+		if (vad_stderr)
+			print_vad(61, 32, false, vad_tx);
+
+		module_event("fvad", "vad", NULL, (struct call*)vad->call,
```

Comment:
Instead NULL pass `call_ua(vad->call)`.

Remove the `struct call*`. `const` for the pointer to call does help here.